Fix the broken index max kernels

argilo opened the pull request:
Fixes https://github.com/gnuradio/volk/issues/359.

The `32fc_index_max_32u` and `32fc_index_max_16u` kernels for `a_avx2` and `u_avx2` currently fail for vector lengths that are not a multiple of 8, due to several bugs in the calculations for the tail end of the vector.

To solve the problem, I've replaced the vectorized tail-end processing with ordinary loops, similar to what's done in the `32f_index_max` kernels.

I also did a bunch of cleanup in some separate commits.

I tested this code with vector lengths 0-32 as follows:

`for l in {0..32}; do for i in {1..50}; do apps/volk_profile -v $l -i 1 -R 32fc_index_max_16u | grep -B 1 fail; apps/volk_profile -v $l -i 1 -R 32fc_index_max_32u | grep -B 1 fail; done; done`

After these changes, all errors are gone.

argilo commented:
It may still be possible to squeeze a bit more performance out of the code that processes the tail end, but for now it seems better to have slow code that works than fast code that doesn't. :smile:

jdemel commented on kernels/volk/volk_32fc_index_max_16u.h:
```diff
@@ -141,104 +138,33 @@ volk_32fc_index_max_16u_a_avx2(uint16_t* target, lv_32fc_t* src0,
+      target[0] = i;
+      sq_dist = new_sq_dist;
+    }
+  }
```
I really prefer this solution over the old one. Since we generally aim for large vectors, this small performance bump that might be introduced here, is probably neglectable.

Besides, this code should be copypasta from the generic kernel. Could you just call the generic kernel on the tail instead?

argilo replied:
> Could you just call the generic kernel on the tail instead?

Not easily, because the kernel only returns the index and not the maximum value itself (which would then need to be recomputed to check whether it's larger than the beginning of the array). That's why I went with the copy & paste approach. The code is actually adapted from the `32f_index_max` AVX kernels, but I'll see if I can make it more like the `32fc_index_max` generic kernels.

jdemel replied:
You could write a function that also returns the maximum value. Then just call that function in the generic kernel.

jdemel commented on kernels/volk/volk_32fc_index_max_16u.h:
```diff
@@ -263,21 +186,20 @@ volk_32fc_index_max_16u_a_sse3(uint16_t* target, lv_32fc_t* src0,
   int i = 0;
 
-  xmm8 = _mm_set_epi32(3, 2, 1, 0);//remember the crazy reverse order!
+  xmm8 = _mm_set_epi32(3, 2, 1, 0);
```
There should be a `_mm_setr_epi32` intrinsic, which hides this reverse order insanity.

argilo replied:
True, I'll add add another commit to change that.

jdemel reviewed (changes requested):
I didn't comment on all code occasions but I assume it would be helpful to update the code as suggested.

Also, you can update the copyright year. `-2020`.

Besides, this PR is really difficult to review because it rewrites so much of these kernels. Still, I like the effort to make it more verbose.

argilo commented:
> Besides, this PR is really difficult to review because it rewrites so much of these kernels.

It may be easier to review the commits individually. I tried to keep the various changes separate to simplify review.

argilo commented:
I made the requested changes. The tail processing code should now be closer to the generic version, and I switched from `set` intrinsics to `setr` in a separate commit.

jdemel commented:
@argilo I just noticed, the copyright year is not updated yet. That's something that should definitely go into this PR.

argilo commented:
Oops, I forgot that. I won't be able to get to it until tonight.

michaelld commented on kernels/volk/volk_32fc_index_max_16u.h:
```diff
@@ -141,104 +140,27 @@ volk_32fc_index_max_16u_a_avx2(uint16_t* target, lv_32fc_t* src0,
-  target[0] = (holderf.f[7] > sq_dist) ? holderi.i[7] : target[0];
-  sq_dist = (holderf.f[7] > sq_dist) ? holderf.f[7] : sq_dist;
-
+    index = sq_dist > max ? i : index;
```
Because there are 2 things being set, I think this would be more clear as:
```
if (sq_dist > max) {
  index = i;
  max = sq_dist;
}
```

michaelld commented on kernels/volk/volk_32fc_index_max_16u.h:
```diff
@@ -492,77 +381,27 @@ volk_32fc_index_max_16u_u_avx2(uint16_t* target, lv_32fc_t* src0,
+    sq_dist = lv_creal(src0[0]) * lv_creal(src0[0]) + lv_cimag(src0[0]) * lv_cimag(src0[0]);
 
-      xmm8 = _mm256_add_epi32(xmm8, xmm10);
+    index = sq_dist > max ? i : index;
```
ditto my comment above here

michaelld commented on kernels/volk/volk_32fc_index_max_32u.h:
```diff
@@ -131,83 +133,33 @@ volk_32fc_index_max_32u_a_avx2(uint32_t* target, lv_32fc_t* src0,
+    sq_dist = lv_creal(src0[0]) * lv_creal(src0[0]) + lv_cimag(src0[0]) * lv_cimag(src0[0]);
 
-    xmm8 = _mm256_add_epi32(xmm8, xmm10);
+    index = sq_dist > max ? i : index;
```
ditto here

michaelld commented on kernels/volk/volk_32fc_index_max_32u.h:
```diff
@@ -445,76 +369,26 @@ volk_32fc_index_max_32u_u_avx2(uint32_t* target, lv_32fc_t* src0,
+    sq_dist = lv_creal(src0[0]) * lv_creal(src0[0]) + lv_cimag(src0[0]) * lv_cimag(src0[0]);
 
-    xmm8 = _mm256_add_epi32(xmm8, xmm10);
+    index = sq_dist > max ? i : index;
```
ditto here

michaelld reviewed (changes requested):
LGTM overall; good to get these kernels fixed! Code reads well & makes sense. Consider my change request ... go with it or not; I think it makes cleans up the code just a touch, but I'm good either way really. Once the Copyright date is updated we can merge IMHO.

argilo commented:
I added a commit with copyright date changes (including the years that had changes according to `git log`) and another commit with the changes suggested by @michaelld. I also updated the generic kernels to match, since that code was copied from them. To make sure I didn't break anything, I first updated the AVX2 kernels, then ran the tests, then updated the generic kernels, and ran tests again.

michaelld commented on kernels/volk/volk_32fc_index_max_16u.h:
```diff
@@ -1,6 +1,6 @@
 /* -*- c++ -*- */
 /*
- * Copyright 2012, 2014 Free Software Foundation, Inc.
+ * Copyright 2012, 2014-2016, 2018-2020 Free Software Foundation, Inc.
```
LOL I'm getting sloppy and just moving to 2012-2020 ... I've seen other projects doing this, too, and it kinda makes sense ... Let's go with your version for now!

michaelld reviewed (approved):
LGTM; thx!

jdemel commented on kernels/volk/volk_32fc_index_max_32u.h:
```diff
@@ -385,12 +313,12 @@ static inline void
 #define INCLUDED_volk_32fc_index_max_32u_u_H
```
I suggest to simplify this include guard. i.e. only one include guard for all kernel versions.

jdemel commented on kernels/volk/volk_32fc_index_max_32u.h:
```diff
@@ -445,76 +375,29 @@ volk_32fc_index_max_32u_u_avx2(uint32_t* target, lv_32fc_t* src0,
+    {
+      index = holderi.i[i];
+      max = holderf.f[i];
+    }
```
I'm curious about formatting here. I'd start braces on a new line for functions and at the end of a line for `for` and `if`. I thought that would be the clang-format way.

argilo replied:
I rebased (to pick up the recent clang-format commit) and ran clang-format on this branch, so the formatting should now be fixed.

jdemel reviewed (approved):
LGTM. Thanks for your patience and your work!

argilo committed:
Fix the broken index max kernels

jdemel commented:
2 approvals, CI passes, merging.
